### ISSUE TITLE
按照你在lenet里面加入stn的方法加入到了我的网络里面，但是训练却不收敛

### ISSUE THREAD

**Issue by December-boy** [Dive-frank/caffe_stn#1]
name: "base_stn"

layer {
  name: "data"
  type: "ImageData"
  top: "data"
  top: "label"
  transform_param {
    mirror: false
    mean_value: 103.939
    mean_value: 116.779
    mean_value: 123.68
  }
  image_data_param {
    source: "external/exp/datalists/${dataset}/${subset}.txt"
    root_folder: ""
    batch_size: 100
    new_height: 96
    new_width: 96
    shuffle: false
  }
}

########################STN########################

layer {
  name: "loc_conv1"
  type: "Convolution"
  bottom: "data"
  top: "loc_conv1"
  convolution_param {
    num_output: 20
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "loc_pool1"
  type: "Pooling"
  bottom: "loc_conv1"
  top: "loc_pool1"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "loc_relu1"
  type: "ReLU"
  bottom: "loc_pool1"
  top: "loc_pool1"
}
layer {
  name: "loc_conv2"
  type: "Convolution"
  bottom: "loc_pool1"
  top: "loc_conv2"
  convolution_param {
    num_output: 20
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "loc_pool2"
  type: "Pooling"
  bottom: "loc_conv2"
  top: "loc_pool2"
  pooling_param {
    pool: MAX
    kernel_size: 2
    stride: 2
  }
}
layer {
  name: "loc_relu2"
  type: "ReLU"
  bottom: "loc_pool2"
  top: "loc_pool2"
}
layer {
  name: "loc_ip1"
  type: "InnerProduct"
  bottom: "loc_pool2"
  top: "loc_ip1"
  inner_product_param {
    num_output: 20
    weight_filler {
      type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "loc_relu3"
  type: "ReLU"
  bottom: "loc_ip1"
  top: "loc_ip1"
}
layer {
  name: "loc_reg"
  type: "InnerProduct"
  bottom: "loc_ip1"
  top: "theta"
  inner_product_param {
    num_output: 6
    weight_filler {
     type: "xavier"
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "st_layer"
  type: "SpatialTransformer"
  bottom: "data"
  bottom: "theta"
  top: "st_output"
  st_param {
    to_compute_dU: false
}
}

######################### conv1 #########################

layer {
  name: "conv1"
  type: "Convolution"
  bottom: "st_output"
  top: "conv1"
  param {
    lr_mult: 1
    decay_mult: 1
  }
  convolution_param {
    num_output: 32
    bias_term: false
    pad: 1
    kernel_size: 3
    stride: 1
    weight_filler {
      type: "xavier"
    }
  }
}

layer {
  name: "conv1_bn"
  type: "BatchNorm"
  bottom: "conv1"
  top: "conv1"
  batch_norm_param {
    use_global_stats: false
  }

**Comment by xiongdemao** [comment]
我的也不收敛，尴尬了

**Comment by HongChen1995** [comment]
问下你们有在windows平台下编译成功的吗？

**Comment by chl916185** [comment]
@xiongdemao ，你的结果怎么样啊
I0710 19:57:47.256626 13014 solver.cpp:414]     Test net output #0: accuracy = 0.9852
I0710 19:57:47.256660 13014 solver.cpp:414]     Test net output #1: loss_cls = 0.0877643 (* 1 = 0.0877643 loss)
I0710 19:57:47.263000 13014 solver.cpp:239] Iteration 167500 (116.78 iter/s, 4.28154s/500 iters), loss = 0.00169689
I0710 19:57:47.263036 13014 solver.cpp:258]     Train net output #0: accuracy = 1
I0710 19:57:47.263059 13014 solver.cpp:258]     Train net output #1: loss_cls = 0.00169646 (* 1 = 0.00169646 loss)
I0710 19:57:47.263067 13014 sgd_solver.cpp:112] Iteration 167500, lr = 0.000213821
I0710 19:57:49.664582 13019 data_layer.cpp:73] Restarting data prefetching from start.
I0710 19:57:51.179571 13014 solver.cpp:347] Iteration 168000, Testing net (#0)
I0710 19:57:51.179602 13014 net.cpp:676] Ignoring source layer accuracy-train
I0710 19:57:51.530228 13020 data_layer.cpp:73] Restarting data prefetching from start.
I0710 19:57:51.544438 13014 solver.cpp:414]     Test net output #0: accuracy = 0.9848
I0710 19:57:51.544466 13014 solver.cpp:414]     Test net output #1: loss_cls = 0.0899908 (* 1 = 0.0899908 loss)
I0710 19:57:51.550964 13014 solver.cpp:239] Iteration 168000 (116.605 iter/s, 4.28799s/500 iters), loss = 0.000439531
I0710 19:57:51.550982 13014 solver.cpp:258]     Train net output #0: accuracy = 1
I0710 19:57:51.550989 13014 solver.cpp:258]     Train net output #1: loss_cls = 0.000439116 (* 1 = 0.000439116 loss)

**Comment by December-boy** [comment]
最终是收敛了，但是效果看起来不是特别好，我也找不到原因，可能是加的地方不对
